// uvcanvas/dist/components/slides/default_components/index.js
"use client"
import React from 'react';

var styles = {"slides":"uvc__slides","cover":"uvc__cover","cover_overlay":"uvc__cover_overlay","ul":"uvc__ul","ol":"uvc__ol","li":"uvc__li"};

const defaultComponents = {
    h1: ({ style, ...props }) => (React.createElement("h1", { style: {
            fontSize: "2em",
            fontWeight: "bold",
            width: "100%",
            borderBottom: "calc(0.25 * var(--svw)) rgb(255 255 255 / 0.5) solid",
            paddingBottom: "calc(1 * var(--svw))",
            marginBottom: "calc(2 * var(--svw))",
            lineHeight: "1.5em",
            ...style,
        }, ...props })),
    li: (props) => (React.createElement("li", { className: styles.li, ...props })),
    ol: (props) => (React.createElement("ol", { className: styles.ol, ...props })),
    ul: (props) => (React.createElement("ul", { className: styles.ul, ...props })),
    Cover: ({ style, className, children, ...props }) => {
        return (React.createElement("div", { className: styles.cover + " " + className, ...props },
            React.createElement("h1", null, props.title),
            props.subtitle && (React.createElement("h2", { style: {} }, props.subtitle)),
            React.createElement("div", { className: styles.cover_overlay },
                React.createElement("header", null, props.header),
                React.createElement("footer", null, props.footer))));
    },
    Background: ({ style, className, children, ...props }) => {
        return (React.createElement("div", { style: {
                background: "black",
                position: "absolute",
                left: "0",
                top: "0",
                width: "100%",
                height: "100%",
                zIndex: "-1",
            }, ...props },
            React.createElement("div", { style: {
                    width: "100%",
                    height: "100%",
                    ...style,
                }, className: className }, children)));
    },
};

export { defaultComponents };
