// uvcanvas/dist/index.js
"use client"
export { Lumiflex } from './components/lumiflex/index.js';
export { Zenitho } from './components/zenitho/index.js';
export { Novatrix } from './components/novatrix/index.js';
export { Velustro } from './components/velustro/index.js';
export { Opulento } from './components/opulento/index.js';
export { Tranquiluxe } from './components/tranquiluxe/index.js';
export { Venturo } from './components/venturo/index.js';
export { Xenon } from './components/xenon/index.js';
export { Slides } from './components/slides/index.js';
export { defaultComponents } from './components/slides/default_components/index.js';
import { evaluateSync } from '@mdx-js/mdx';
import * as runtime from 'react/jsx-runtime';
export { SlidesActionButton } from './components/slides/action_button/index.js';
import 'ogl';
import 'react';
import 'react-hotkeys-hook';
import '@radix-ui/react-aspect-ratio';
import 'react-resize-detector';
import 'jotai';
import './state.js';

function sliceMdxString(mdx) {
    const rawSlices = mdx.split(/^---$/gm).map((mdx) => mdx.trim());
    const slices = rawSlices.map((mdx) => evaluateSync(mdx, runtime));
    return slices;
}

export { sliceMdxString };
